### ISSUE TITLE
New Rule: Text nodes have minimal contrast (afw4f7)

### ISSUE THREAD

**Issue by WilcoFiers** [act-rules/act-rules.github.io#571]
---
id: afw4f7
name: Text nodes have minimal contrast
rule_type: atomic

description: |
  This rule checks that text nodes have minimal color contrast with its background

accessibility_requirements:
  wcag20:1.4.3: # Contrast (Minimum)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling

authors:
  - Wilco Fiers
  - Kasper Isager
---

## Applicability

Any text node that is [visible](#visible) and a child of an HTML element, except the text node is a descendent of an element that:

- Has a [semantic role]() that inherits from [widget]()
- Is used in the accessible name of a widget that is [disabled]() (i.e. It is part of the label)
- Has a [semantic role]() of [group]() with that is [disabled]()

**Note**: When the text color is the same as the background color, the element is not visible, and so is not applicable.

## Expectation

For each test target, the [highest possible contrast]() between the [background colors]() and [foreground colors]() is at least 4.5:1 (or 3.0:1 for [larger scale text]()), except if the content of the text node:

- is [decorative](#decorative), or
- doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s)

**Note**: Passing this rule does not mean that the text node has sufficient color contrast. White text on a black and white image passes this rule, whether or not that text is legible depends on the how much of the white words are positioned on the black parts of the image. This requires further testing.

## Assumptions

Success criterion 1.4.3 has exceptions for "incidental" text, which includes inactive user interface components and decorative texts. The rule assumes that text nodes should be ignored are hidden from assistive technologies. If this isn't the case, the rule may produce incorrect results.

TODO: The text node is part of a logo or brand name

## Accessibility support

_There are no major accessibility support issues known for this rule._

## Background

- (e.g. WCAG Techniques or links with background information mentioned in Applicability, Expectations or Assumptions)

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

#### Passed Example X

overlays

#### Passed Example X

blurry decorative text

#### Passed Example X

Text nodes decorating an image

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...

----

### Foreground Colors of test nodes

All the colors of the pixels of a [text node]() that change when the CSS `color` property is changed, except for the colors of [aliased pixels]().

TODO: What if a text only has aliased pixels, is that possible?

### Background Colors of test nodes

All colors not part of the [foreground] in a bounding box around each [visible] character 

TODO, must account for shadow https://codepen.io/pen/?editors=0100

Except for aliased pixels

### Highest Possible Contrast

Todo

### Larger Scale Text

with at least 18 point or 14 point bold or font size that would yield equivalent size for Chinese, Japanese and Korean (CJK) fonts

**TODO**: Deal with "bold" fonts
**TODO**: How to handle CJK fonts?

**Comment by dd8** [comment]
The applicability should not exclude `aria-hidden=true` since this affects lots of users groups who don't use AT (e.g. color blind users)

There's at least one support issue - the following code fails the contrast test in Firefox, but passes in all other browsers:

```
<!doctype html>
<html lang="en">
<head>
    <title>Test</title>
    <style>
    body {
        color: #000;
        background-color: #fff;
    }

    input {
        color: #ddd;
        background-color: #ccc;
    }

    /* :read-only and :read-write only supported in Blink/WebKit/Edge/IE11 - not supported in Firefox */
    input:read-only {
        color: #111;
        background-color: #eee;
    }

    input:read-write {
        color: #000;
        background-color: #fff;
    }
    </style>
</head>
<body>
    <label>Read only:<input type='text' readonly value='read only text'></label>
    <label>Read write:<input type='text'  value='read write text'></label>
</body>
</html>
```

**Comment by WilcoFiers** [comment]
Good points Mark.

> The applicability should not exclude aria-hidden=true since low contrast affects lots of users groups who don't use AT (e.g. color blind users)

Perhaps you're right. I'm definitely not trying to say that everything in aria-hidden doesn't need to have contrast. The reason I put it in there is to avoid some false positives that I've seen, where decorative text was given aria-hidden="true". Alternatively what we could do is add an exception in the expectation to deal with that, but that makes it rather difficult to automate. I'm not even sure about the logo's exception. Our test tool certainly doesn't take that into account. I'm not sure how we could ever do that.

> There's at least one support issue - the following code fails the contrast test in Firefox and IE11, but passes in all other browsers:

I'm not sure if browser support for certain CSS can be considered accessibility support. Looking up the definition in WCAG it says "supported by users' assistive technologies as well as the accessibility features in browsers and other user agents". CSS isn't an accessibility feature in a browser, I would say that this doesn't apply.

It's kinda weird though. It seems obvious something like that should fail WCAG, but I don't think there's anything in WCAG that actually says it does. Does it even make sense to ask if CSS `:read-only` is accessibility supported? It's not an accessibility feature, so the question doesn't make sense.

**Comment by dd8** [comment]
Yeah, I don't think it fits the accessibility support definition, but there's a few SCs like this one that apply where no AT is involved. Due due to inconsistent standards support or the author using conditional features like @supports and @media you can get failures that affect some, but not all, users.

For example:
```
    input {
        color: #ddd;
        background-color: #ccc;
    }
    /* contrast fail on screens less than 800px wide = mobiles and tablets in portrait mode */
    @media ( min-width: 800px ) {
      input:read-only {
        color: #111;
        background-color: #eee;
      }
    }
```

Is it a fail or a pass in this case?

If it's a pass then consider what happens when the CSS rule displays a flashing animation failing 2.3.1 instead of causing a contrast issue. That means you have content that passes but can cause seizures for users of certain devices.

**Comment by dd8** [comment]
I started a thread on w3-wai-ig on this
http://lists.w3.org/Archives/Public/w3c-wai-ig/2019AprJun/0085.html

There conclusions on this were:

1) Contrast above threshold in some browsers, but below in others due to different levels of CSS support in different browsers (e.g. the :read-write example in the comment above, also covers @supports)
**Pass** since the selectors used are a 'relied upon' technology.
(but human evaluators may come to a different conclusion depending on which browser(s) they test with)

2) Contrast above threshold at some screen sizes, but below in others due CSS media queries 
**Fail** since WCAG 2.1 has a requirement for SCs to pass at different screen sizes https://www.w3.org/TR/WCAG21/#cc2
(but would this be a pass in WCAG 2.0?)

3) Contrast above threshold in some browsers, but below in others due to browsers differences in handling CSS parsing errors 

**Pass** or **Fail** different auditors may come to different conclusions - and practically speaking will depend on which browser(s) they use for testing 
> Patrick H. Lauke wrote:
>I'm  sure in general everybody would agree that "it's a problem", 
>the gray  area is whether it's a hard fail or not...

The problem here is a detected contrast issue could be WCAG pass or fail, depending on the cause :-(
Some wording in the applicability like the following might avoid the issue:
- CSS used has no parsing errors
- browsers support the 'relied upon' technologies (but might be hard for anyone to figure out what these are)

**Comment by Jym77** [comment]
Closed by #833.
